sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/analysis/CheckAnalysis.scala

```diff
@@ -110,6 +110,31 @@ trait CheckAnalysis {
               s"filter expression '${f.condition.sql}' " +
                 s"of type ${f.condition.dataType.simpleString} is not a boolean.")
 
+          case f @ Filter(condition, child) =>
+            // Make sure no correlated predicate is in an OUTER join, because this could change the
+            // semantics of the join.
+            lazy val attributes: Set[Expression] = child.output.toSet
+            def checkCorrelatedPredicates(p: PredicateSubquery): Unit = p.query.foreach {
+              case j @ Join(left, right, jt, _) if jt != Inner =>
+                j.transformAllExpressions {
+                  case e if attributes.contains(e) =>
+                    failAnalysis(s"Accessing outer query column is not allowed in outer joins: $e")
```

Comment:
Outer join is not the only one that change the meaning of an attribute (for example, Union), so we can't access the outer attributes below these plan (outer side of outer join, the children of Union (other than the first one). We can create a TODO for these.